test/python/circuit/test_twirling.py

```diff
@@ -0,0 +1,212 @@
+        qc = QuantumCircuit(2)
+        qc.append(gate(), (0, 1))
+        res = twirl_circuit(qc, gate, seed=424242, num_twirls=1000)
+        for twirled_circuit in res:
+            np.testing.assert_allclose(
+                Operator(qc), Operator(twirled_circuit), err_msg=f"gate: {gate} not equiv to\n{res}"
+            )
+            self.assertNotEqual(twirled_circuit, qc)
+
+    def test_invalid_gate(self):
+        """Test an error is raised with a non-standard gate."""
+
+        class MyGate(Gate):
+            """Custom gate."""
+
+            def __init__(self):
+                super().__init__("custom", num_qubits=2, params=[])
+
+        qc = QuantumCircuit(2)
+        qc.append(MyGate(), (0, 1))
+
+        with self.assertRaises(QiskitError):
+            twirl_circuit(qc, twirling_gate=MyGate())
+
+    def test_custom_standard_gate(self):
+        """Test an error is raised with an unsupported standard gate."""
+        qc = QuantumCircuit(2)
+        qc.swap(0, 1)
+        res = twirl_circuit(qc, twirling_gate=SwapGate())
+        np.testing.assert_allclose(
+            Operator(qc), Operator(res), err_msg=f"gate: {qc} not equiv to\n{res}"
+        )
+        self.assertNotEqual(qc, res)
+
+    def test_invalid_string(self):
+        """Test an error is raised with an unsupported standard gate."""
+        qc = QuantumCircuit(2)
+        qc.swap(0, 1)
+        with self.assertRaises(QiskitError):
+            twirl_circuit(qc, twirling_gate="swap")
+
+    def test_invalid_str_entry_in_list(self):
+        """Test an error is raised with an unsupported string gate in list."""
+        qc = QuantumCircuit(2)
+        qc.swap(0, 1)
+        with self.assertRaises(QiskitError):
+            twirl_circuit(qc, twirling_gate=[CXGate, "swap"])
+
+    def test_invalid_class_entry_in_list(self):
+        """Test an error is raised with an unsupported string gate in list."""
+        qc = QuantumCircuit(2)
+        qc.swap(0, 1)
+        res = twirl_circuit(qc, twirling_gate=[SwapGate(), "cx"])
+        np.testing.assert_allclose(
+            Operator(qc), Operator(res), err_msg=f"gate: {qc} not equiv to\n{res}"
+        )
+        self.assertNotEqual(qc, res)
+
+    @ddt.data(CXGate, ECRGate, CZGate, iSwapGate)
+    def test_full_circuit(self, gate):
+        """Test a circuit with a random assortment of gates."""
+        qc = random_circuit(5, 25, seed=12345678942)
+        qc.append(PermutationGate([1, 2, 0]), [0, 1, 2])
+        res = twirl_circuit(qc)
+        np.testing.assert_allclose(
+            Operator(qc), Operator(res), err_msg=f"gate: {gate} not equiv to\n{res}"
+        )
+
+    @ddt.data(CXGate, ECRGate, CZGate, iSwapGate)
+    def test_control_flow(self, gate):
+        """Test we twirl inside control flow blocks."""
+        qc = QuantumCircuit(2, 1)
+        with qc.if_test((qc.clbits[0], 0)):
+            qc.append(gate(), [0, 1])
+        res = twirl_circuit(qc)
```

Comment:
This should maybe also check that the circuits are not equal (maybe with a seed to avoid randomly hitting the identity-twirl case)